### ISSUE TITLE
Make Hypha (only apply part) into a python package on pypi

### ISSUE THREAD

**Issue by frjo** [OpenTechFund/hypha#2144]
1. OTF and Reset will keep using the public part. Should we make that a separate python package or easier to let them maintain there on installs?
2. What needs to be done to have the public and apply part for OTF and Reset  work together as now when the apply is a python package?
3. API that lists open rounds and labs. Something that can be used to list them when the public side is an Wordpress install or something else.
4. Other API that would be useful for whatever public sites other Hypha users will have?
5. How mush trouble will it be to migrate current Hypha installs to the Hypha apply package? Can the DB stay intact?
6. How much work will this be? (very ruff numbers)

**Comment by sks444** [comment]
@frjo, good initiative. This could make it easy for new organizations to start using hypha workflows.

We should consider this as a long term task and divide it into smaller tasks. The plan is to move one app at time from the [apply](https://github.com/OpenTechFund/hypha/tree/master/hypha/apply) directory.

E.g. First move the [funds](https://github.com/OpenTechFund/hypha/tree/master/hypha/apply/funds) app to the package. Release and use it in Hypha/Reset. Then move reviews app to the package for its next release. Similarly others.

Moving one app at a time will decrease complexity. And also it doesn't break any development workflow as we keep on using the current apps until it is permanently moved to the package. This way we also get to see/experiment on how much time it is taking to move one app. And we get time to think/decide on what to keep/what to move.

>OTF and Reset will keep using the public part. Should we make that a separate python package or easier to let them maintain there on installs?

It would be easier to have public part maintained by the organization. As it would different for every org. Whereas apply part(workflow) would be same for all the orgs.


>What needs to be done to have the public and apply part for OTF and Reset work together as now when the apply is a python package?

We need to fix our imports of apply modules to use it from the package instead of using it from apply directory in hypha. 

>API that lists open rounds and labs. Something that can be used to list them when the public side is an Wordpress install or something else.

An api could be build for such scenario but we should target this as one of the last tasks. As to use it in hypha/reset we just need to import the correct modules and everything will be same as it is now.

>How mush trouble will it be to migrate current Hypha installs to the Hypha apply package? Can the DB stay intact?

Most probably, We won't have to change anything to db as all the migration files that we have will be the same, it is just that they will be coming from the installed package instead of our current apps. So no migrate to db will be done. But I am not sure yet. As this will be in a new repo, migration files could change, in that case we might need to write custom migrations to have similar table names..etc.

>How much work will this be?

This seems like a lot of work. But once we move one of the apps to the package and start using it in Hypha/Reset. We would have more clear idea about the others. It doesn't seem like a task involving much complexity through-out its completion but yet could be time consuming. Initially setting up the package/releasing it to PyPi could also take time but once we get familiar with process, it would become easy.

To conclude, it would be kind of similar to managing a package like wagtail with a lot of features built on top of it. Which provides all of our apply app functionalities.

**Comment by frjo** [comment]
@sks444 Glad to hear you seem to have a handle on this! I like the suggestion of moving app by app. IF we can get started before the ned of the year I would be happy but maybe more likely this will be next year.

**Comment by sks444** [comment]
Okay. Just to mention whenever we start working on this, we would need a new repository for the package.

**Comment by frjo** [comment]
Maybe https://github.com/HyphaApp/hypha

**Comment by Parbhat** [comment]
>On a related note, we are looking into the pip install package for apply app. The goal is to create apply app a separate package and use it as pip install on OTF and reset. Things do not seem that simple to me. As moving it simply will not make it truly reusable and base django project should be configured in the way to use it. I think the goal is not have to create a fork (similar to Reset) but fork has an added advantage to create changes in code directly. How we will make changes in a separate apply package say we will need to add custom fields on Application Submission model? We will have to make abstraction on the code level to allow making custom changes. Like third party packages like Wagtail provides functionality and also allow to override it like using a custom image model. - Slack

**Comment by frjo** [comment]
@Parbhat Have you and your team had time to think about this?

**Comment by Parbhat** [comment]
@frjo Yeah, had a chat with @sks444 about this and this is a major chunk of work for making it truly reusable. If we make it separate package without making abstracts interface for making custom changes then it will be harder to make small updates also.
